source/mod_particles.f90

```diff
@@ -20,55 +18,133 @@ end subroutine part_allocate
 
   real(kind=fPrec), intent(in) :: refEnergy
 
   real(kind=fPrec) e0o, e0fo
-  integer          j
+
+  if(e0 == refEnergy) return
+
+  ! Save previous values
+  e0o    = e0
+  e0fo   = e0f
 
   ! Modify the reference particle
-  e0o  = e0
-  e0fo = e0f
-  if(e0 /= refEnergy) then
-    e0     = refEnergy
-    e0f    = sqrt(e0**2 - nucm0**2)
-    gammar = nucm0/e0
+  e0     = refEnergy
+  e0f    = sqrt(e0**2 - nucm0**2)
+  gammar = nucm0/e0
+
+  ! Also update sigmv with the new beta0 = e0f/e0
+  sigmv(1:napx) = ((e0f*e0o)/(e0fo*e0))*sigmv(1:napx)
+
+  if(e0 <= pieni) then
+    write(lout,"(a)") "PART> ERROR Reference energy ~= 0"
+    call prror(-1)
   end if
 
-  ! Modify the Energy
+  call part_updatePartEnergy(1)
+
+end subroutine part_updateRefEnergy
+
+! ================================================================================================ !
+!  K.N. Sjobak, V.K. Berglyd Olsen, BE-ABP-HSS
+!  Last modified: 2018-11-02
+!  Updates the relevant particle arrays after the particle's energy, momentum or delta has changed.
+! ================================================================================================ !
+subroutine part_updatePartEnergy(refArray)
+
+  use mod_hions
+  use mod_common
+  use mod_commont
+  use mod_commonmn
+  use numerical_constants
+
+  implicit none
+
+  integer, intent(in) :: refArray
+
+  real(kind=fPrec) e0o, e0fo
+  integer          j
+
+  select case(refArray)
+  case(1) ! Update from energy array
+    do j=1, napx
+      ejfv(j) = sqrt(ejv(j)**2 - nucm(j)**2)        ! Momentum [MeV/c]
+      dpsv(j) = (ejfv(j)*(nucm0/nucm(j))-e0f)/e0f   ! Delta_p/p0 = delta
+    end do
+  case(2) ! Update from momentum array
+    do j=1, napx
+      ejv(j)  = sqrt(ejfv(j)**2 + nucm(j)**2)       ! Energy [MeV]
+      dpsv(j) = (ejfv(j)*(nucm0/nucm(j))-e0f)/e0f   ! Delta_p/p0 = delta
+    end do
+  case(3) ! Update from delta array
+    do j=1, napx
+      ejfv(j) = ((nucm(j)/nucm0)*(dpsv(j)+one))*e0f ! Momentum [MeV/c]
+      ejv(j)  = sqrt(ejfv(j)**2 + nucm(j)**2)       ! Energy [MeV]
+    end do
+  case default
+    write(lout,"(a)") "PART> ERROR Internal error in part_updatePartEnergy"
+    call prror(-1)
+  end select
+
+  ! Modify the Energy Dependent Arrays
   do j=1, napx
-    ejfv(j)     = sqrt(ejv(j)**2 - nucm(j)**2)       ! Momentum [MeV/c]
-    dpsv(j)     = (ejfv(j)*(nucm0/nucm(j))-e0f)/e0f  ! Delta_p/p0 = delta
     dpsv1(j)    = (dpsv(j)*c1e3)/(one + dpsv(j))
-    dpd(j)      = one + dpsv(j)
-    dpsq(j)     = sqrt(dpd(j))
+    dpd(j)      = one + dpsv(j)                      ! For thick tracking
+    dpsq(j)     = sqrt(dpd(j))                       ! For thick tracking
     oidpsv(j)   = one/(one + dpsv(j))
     moidpsv(j)  = mtc(j)/(one + dpsv(j))             ! Relative rigidity offset (mod_hions) [MV/c^2]
-    omoidpsv(j) = c1e3*((one-mtc(j))*oidpsv(j))
+    omoidpsv(j) = ((one-mtc(j))*oidpsv(j))*c1e3
     rvv(j)      = (ejv(j)*e0f)/(e0*ejfv(j))          ! Beta_0 / beta(j)
   end do
 
```

Comment:
shouldn't we perform the operations in the same order as done in the version of sixtrack for regular DA studies?
i.e.:
```
   oidpsv(j) = [...]
   moidpsv(j)  = [...]
   omoidpsv(j) = c1e3*((one-mtc(j))*oidpsv(j))
   dpsv1(j)=(dpsv(j)*c1e3)*oidpsv(j)
   rvv(j) = [...]
   dpd(j) = [...]
   dpsq(j) = [...]
```

This would imply also in the upper `select`-`case` statement:
```
case(1)
   ! momentum is fine, though I would use sqrt((ejv(j)-nucm(j))*(ejv(j)+nucm(j))) 
   dpsv(j)=(ejfv(j)/e0f)*(nucm0/nucm(j))-one
case(2)
   ! energy is fine
   dpsv(j)=(ejfv(j)/e0f)*(nucm0/nucm(j))-one
case(3)
   ejfv(j)=e0f*(one+dpsv(j))*(nucm(j)/nucm0)
   ! energy is fine
```

Reply:
The order probably changed because I moved things around. The order of calculation is in any case determined by the brackets, no?

Reply:
ok